src/types.ts

```diff
@@ -20,3 +25,41 @@ export interface Block extends Node, BlockBase {
 export type Message = Linter.LintMessage;
 
 export type RuleType = "problem" | "suggestion" | "layout";
+
+/**
+ * The `SourceCode` interface for Markdown files.
+ */
+export interface IMarkdownSourceCode extends TextSourceCode {
+	/**
+	 * Gets the entire source text split into an array of lines.
+	 * @returns {Array<string>} The source text as an array of lines.
+	 * @public
+	 */
+	get lines(): Array<string>;
+
+	/**
+	 * Gets the source code for the given node.
+	 * @param {object} [node] The AST node to get the text for.
+	 * @param {number} [beforeCount] The number of characters before the node to retrieve.
+	 * @param {number} [afterCount] The number of characters after the node to retrieve.
+	 * @returns {string} The text representing the AST node.
+	 * @public
+	 */
+	getText(node?: object, beforeCount?: number, afterCount?: number): string;
+}
+
+export type MarkdownRuleVisitor = Record<
+	string,
+	((node: Node) => void) | undefined
+>;
```

Comment:
Can we add some of the methods with node names like we do for the JSON plugin?
https://github.com/eslint/json/blob/10882ffe9c39cdd866be51801f9950f4a010cd87/src/types.ts#L58-L84

Reply:
Done in 4e326af917f7421935b54bc3fe34e89fc87b24aa. I've added listeners for all the node types used in our rules plus `Root`. Please, check if there is anything else that should be added.